feat: Adds support for HNS L4WFPProxyPolicy

davinci26 opened the pull request:
Fixes #1002

Allow the cni plugin to marshall and apply L4WFPProxyPolicy
to Windows endpoints.

Tested on Kubernetes v1.19 with AKS-engine and docker runtime

Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

<!-- Thank you for helping Azure Container Networking with a pull request!
Use conventional commit messages, such as
  feat: add a knob to the frobnitz
or
  fix: repair hole in wumpus
And read this for faster PR reviews: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#best-practices-for-faster-reviews -->

**Reason for Change**:
<!-- What does this PR improve or fix in Azure Container Networking? -->


**Issue Fixed**:
<!-- If this PR fixes GitHub issue 1234, add "Fixes #1234" to the next line. -->


**Requirements**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR. -->


- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  <!-- Common commit types:
        build: Build 🏭
        chore: Maintenance 🔧
        ci: Continuous Integration 💜
        docs: Documentation 📘
        feat: Features 🌈
        fix: Bug Fixes 🐞
        perf: Performance Improvements 🚀
        refactor: Code Refactoring 💎
        revert: Revert Change ◀️
        style: Code Style 🎶
        security: Security Fix 🛡️
        test: Testing 💚 -->
- [ ] includes documentation
- [ ] adds unit tests


**Notes**:

davinci26 commented:
Some comments:

I see that the other policies do not have unit tests so I did not add any tests but I am happy to add tests as needed. Similar for the documentation.

I have verified this change with Kubernetes + docker, let me know if you want me to verify this with Kubernetes + containerd as well.

In general I am looking for your guidance and I am happy to do any validation/changes needed to get this PR merged

rbtr commented:
first pass lgtm so I kicked off the CI
@davinci26 actually this week we are doing a quality week focused on unit tests, if you are up to UT this scenario that would be much appreciated 😊 
k8s x containerd validation would be good. does this functionality deserve a full e2e scenario?

rbtr commented:
/azp run

davinci26 commented:
> if you are up to UT this scenario that would be much appreciated 

I can try to do that.

> k8s x containerd validation would be good

I will validate that today

> does this functionality deserve a full e2e scenario?

I dont think this is is needed as this is mostly a dev solution.

Is the errors in the CI due to my change?

rbtr commented:
negative, it's a flaky test 👍🏼

davinci26 commented:
@rbtr as I was working through the containerd + kubernetes integration I realized that I need to update the version of `github.com/Microsoft/hcsshim` to `v0.8.22` from `v0.8.10` to pick up the api change.

Making this change is making the diff 300+ files where all the files are in the vendor path. Does this seem correct?

Also I have added unit tests but you won't be able to run them in CI because they require `GOOS=windows go test -v ./network/policy` to be invoked but I think you have a draft PR for that so they will become available

rbtr commented:
hey @matmerr maybe you could bump your https://github.com/Azure/azure-container-networking/pull/961 to 0.8.22 and we can merge that before this?

davinci26 commented:
Ok lets not merge this until #961 goes in (I still need to push my tests, I have them locally)

matmerr commented:
> hey @matmerr maybe you could bump your #961 to 0.8.22 and we can merge that before this?

would like to have @ashvindeodhar's eyes on this as well as #961

ashvindeodhar reviewed (approved):
lgtm

davinci26 commented:
Rebased and added a unit test for my feature as discussed. Please review if the test is in the correct direction, currently I invoke it with `GOOS=windows go test -v ./network/policy` but this mean that the test will not be invoked in the CI

rbtr commented:
/azp run

davinci26 commented:
@ashvindeodhar can you please take a look at this PR

davinci26 commented:
Am I going to be too annoying if I also ask to carve out a release? I want to ingest this change into aks-engine

rbtr commented:
You want a new Windows CNI release, right?

davinci26 commented:
I want to push a pr similar to this https://github.com/Azure/aks-engine/pull/4656/files so I would need release 1.4.13

rbtr commented:
@davinci26 v1.4.13 is released, includes your change
